File: android/src/main/kotlin/dev/pam/scanner/RecentScans.kt
package dev.pam.scanner

internal class RecentScans(private val capacity: Int = 128) {
    private val seen = linkedMapOf<String, Long>()

    init {
        require(capacity > 0)
    }

    fun accept(value: String, now: Long, interval: Long): Boolean {
        if (value.isEmpty()) return false
        seen.entries.removeAll { now - it.value >= interval }
        if (seen.containsKey(value)) return false
        if (seen.size >= capacity) seen.remove(seen.keys.first())
        seen[value] = now
        return true
    }

    fun clear() = seen.clear()
}

File: ios/Sources/RecentScans.swift
import Foundation

final class RecentScans {
    private let capacity: Int
    private var entries: [(value: String, timestamp: Int64)] = []

    init(capacity: Int = 128) {
        precondition(capacity > 0)
        self.capacity = capacity
    }

    func accept(_ value: String, now: Int64, interval: Int64) -> Bool {
        guard !value.isEmpty else { return false }
        entries.removeAll { now - $0.timestamp >= interval }
        guard !entries.contains(where: { $0.value == value }) else { return false }
        if entries.count >= capacity { entries.removeFirst() }
        entries.append((value, now))
        return true
    }

    func clear() { entries.removeAll() }
}
